test/utils/widget_pump.dart

```diff
@@ -9,3 +11,16 @@ Future<void> pumpMaterialScoped(WidgetTester tester, Widget widget) => tester.pu
         ),
       ),
     );
+
+/// Wraps a [pumpMaterialScoped] with a [ProviderScope] that overrides [themeController] and optional [overrides].
+Future<void> pumpThemedProviderScoped(WidgetTester tester, Widget widget, [List<Override> overrides = const []]) =>
+    tester.pumpWidget(
+      ProviderScope(
+        overrides: List.from(overrides)..add(themeController.overrideWithValue(ThemeController())),
+        child: MaterialApp(
+          home: Scaffold(
+            body: widget,
+          ),
+        ),
+      ),
+    );
```

Comment:
```suggestion
/// Wraps a [pumpMaterialScoped] with a [ProviderScope] that optionally uses [overrides].
Future<void> pumpThemedProviderScoped(WidgetTester tester, Widget widget, [List<Override> overrides = const []]) =>
    tester.pumpWidget(
      ProviderScope(
        overrides: overrides,
        child: MaterialApp(
          home: Scaffold(
            body: widget,
          ),
        ),
      ),
    );
```

Make the callers pass what they want.

Also:

> But we are not using it directly, we use it through useTheme which trie to read it from the context scope.

It's that once you give a `ProviderScope`, it knows how to access the `themeController`, because just like in the application, we don't override anything from the theme, we use the provider-provided one.

Have you tried running these tests WITHOUT overriding the `ThemeController`, which is provided by default https://github.com/olmps/memo/blob/9a9a8197fd0f8fe3af0213a408bcc166462aea81/lib/application/theme/theme_controller.dart#L10 ?

Reply:
Now I get it. It works without the `ThemeController` override, just with the normal `ProviderScope.override`. Now sure what I've tested before that didn't work (I think I was removing the entire `ProviderScope`).